File: filter-service/src/main/scala/SimpleMain.scala
import java.util.Properties

import actor.GeofenceActor
import actor.GeofenceActor.GeofenceRequest
import akka.actor.ActorSystem
import akka.http.scaladsl.model.Uri
import akka.pattern.ask
import akka.stream.ActorMaterializer
import akka.util.Timeout
import client.GeofenceClient
import domain.GPSUpdate
import kafka.KafkaJSONDeserializer
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord}
import org.apache.kafka.common.serialization.{StringDeserializer, StringSerializer}
import play.api.libs.json.Json

import scala.collection.JavaConverters._
import scala.concurrent.duration._
import scala.util.{Failure, Success}


object SimpleMain extends App {

  val kafkaTopicIn = "jfall-gps-updates"
  val kafkaTopicOut = "jfall-filtered-updates"
  val geofenceUri = Uri("http://localhost:9000/coordinates")

  implicit val system: ActorSystem = ActorSystem()
  implicit val materializer: ActorMaterializer = ActorMaterializer()

  // create the producer
  val producerProperties = new Properties()
  producerProperties.put("bootstrap.servers", "localhost:9092")
  producerProperties.put("key.serializer", classOf[StringSerializer].getName)
  producerProperties.put("value.serializer", classOf[StringSerializer].getName)
  val producer = new KafkaProducer[String, String](producerProperties)


  // create the consumer
  val consumerProperties = new Properties()
  consumerProperties.put("bootstrap.servers", "localhost:9092")
  consumerProperties.put("group.id", "filter-service-actor")
  val consumer = new KafkaConsumer[String, GPSUpdate](consumerProperties, new StringDeserializer, new KafkaJSONDeserializer[GPSUpdate]())
  consumer.subscribe(List(kafkaTopicIn).asJava)

  import system.dispatcher
  implicit val timeout: Timeout = Timeout(5 seconds)

  // initialize the logic required to filter updates
  val geofenceClient = GeofenceClient(geofenceUri)
  val geofenceActor = system.actorOf(GeofenceActor.props(geofenceClient))

  val kafkaPollTimeout = 1000

  println("Started filter service")

  while (true) {
    val records = consumer.poll(kafkaPollTimeout)
    records.forEach(record => {

      val gpsUpdate = record.value()
      geofenceActor.ask(GeofenceRequest(gpsUpdate)).onComplete {
        case Success(true) =>
          println("in geofence!")
          produceItem(gpsUpdate)
        case Success(false) => println("Not in geofence")
        case Failure(_) => println("Timeout occurred!")
      }

    })
  }

  def produceItem(loc: GPSUpdate) = {
    producer.send(new ProducerRecord[String, String](kafkaTopicOut, Json.toJson(loc).toString))
  }
}

File: java/filterservice/src/main/java/SimpleMain.java
import actor.GeofenceActor;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.dispatch.OnComplete;
import akka.pattern.Patterns;
import akka.util.Timeout;
import client.GeofenceClient;
import domain.GPSUpdate;
import kafka.KafkaJSONDeserializer;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.kafka.common.serialization.StringSerializer;
import scala.concurrent.Future;

import java.time.Duration;
import java.util.Collections;
import java.util.Properties;
import java.util.concurrent.TimeUnit;

public class SimpleMain {

    private static KafkaProducer<String, String> producer;
    private static final String KAFKA_TOPIC_OUT = "jfall-filtered-updates";
    private static final String KAFKA_TOPIC_IN = "jfall-gps-updates";
    private static final String GEOFENCE_URI = "http://localhost:9000/coordinates";

    public static void main(String args[]) {
        ActorSystem system = ActorSystem.create("routes");

        createProducer();
        KafkaConsumer<String, GPSUpdate> consumer = createConsumer();

        final Timeout askTimeout = new Timeout(5, TimeUnit.SECONDS);

        // initialize the logic required to filter updates
        GeofenceClient geofenceClient = new GeofenceClient(GEOFENCE_URI, system);
        ActorRef geofenceActor = system.actorOf(GeofenceActor.props(geofenceClient));

        Duration kafkaPollTimeout = Duration.ofSeconds(1);

        while (true) {
            ConsumerRecords<String, GPSUpdate> records = consumer.poll(kafkaPollTimeout);
            for (ConsumerRecord<String, GPSUpdate> record : records) {
                GPSUpdate gpsUpdate = record.value();
                Future<Object> futureResponse = Patterns.ask(geofenceActor, new GeofenceActor.GeofenceRequest(gpsUpdate), askTimeout);

                futureResponse.onComplete(new OnComplete<Object>() {
                    @Override
                    public void onComplete(Throwable failure, Object success) {
                        if (failure != null) {
                            System.out.println("Timeout occurred!");
                        } else {
                            if((Boolean) success) {
                                System.out.println("In geofence!");
                                produceItem(gpsUpdate);
                            } else {
                                System.out.println("Not in geofence!");
                            }
                        }
                    }
                }, system.dispatcher());
            }
        }
    }

    private static KafkaConsumer<String, GPSUpdate> createConsumer() {
        // create the consumer
        Properties consumerProperties = new Properties();
        consumerProperties.put("bootstrap.servers", "localhost:9092");
        consumerProperties.put("group.id", "filter-service-actor");
        KafkaConsumer<String, GPSUpdate> consumer = new KafkaConsumer<>(consumerProperties, new StringDeserializer(), new KafkaJSONDeserializer<>(GPSUpdate.class));
        consumer.subscribe(Collections.singleton(KAFKA_TOPIC_IN));
        return consumer;
    }

    private static void createProducer() {
        // create the producer
        Properties producerProperties = new Properties();
        producerProperties.put("bootstrap.servers", "localhost:9092");
        producerProperties.put("key.serializer", StringSerializer.class);
        producerProperties.put("value.serializer", StringSerializer.class);
        producer = new KafkaProducer<>(producerProperties);
    }

    private static void produceItem(GPSUpdate location) {
        producer.send(new ProducerRecord<>(KAFKA_TOPIC_OUT, location.toJson()));
    }
}
